File: src/main/java/br/unifor/costify/application/factory/RecipeFactory.java
package br.unifor.costify.application.factory;

import br.unifor.costify.domain.contracts.IdGenerator;
import br.unifor.costify.domain.entity.Recipe;
import br.unifor.costify.domain.valueobject.Id;
import br.unifor.costify.domain.valueobject.Money;
import br.unifor.costify.domain.valueobject.RecipeIngredient;
import org.springframework.stereotype.Component;
import java.util.List;

@Component
public class RecipeFactory {
  private final IdGenerator idGenerator;

  public RecipeFactory(IdGenerator idGenerator) {
    this.idGenerator = idGenerator;
  }

  public Recipe create(String name, List<RecipeIngredient> ingredients, Money totalCost) {
    return new Recipe(idGenerator, name, ingredients, totalCost);
  }

  public Recipe create(Id id, String name, List<RecipeIngredient> ingredients, Money totalCost) {
    return new Recipe(id, name, ingredients, totalCost);
  }
}

File: internal/application/factory/recipe_factory.go
package factory

import (
	"github.com/vini/costify-go/internal/domain/contract"
	"github.com/vini/costify-go/internal/domain/entity"
	"github.com/vini/costify-go/internal/domain/valueobject"
)

// RecipeFactory provides factory methods for creating recipes
type RecipeFactory struct {
	idGenerator contract.IdGenerator
}

// NewRecipeFactory creates a new RecipeFactory
func NewRecipeFactory(idGenerator contract.IdGenerator) *RecipeFactory {
	return &RecipeFactory{
		idGenerator: idGenerator,
	}
}

// Create creates a recipe with generated ID
func (f *RecipeFactory) Create(
	name string,
	ingredients []valueobject.RecipeIngredient,
	totalCost valueobject.Money,
) (*entity.Recipe, error) {
	return entity.NewRecipeWithGenerator(
		f.idGenerator,
		name,
		ingredients,
		totalCost,
	)
}

// CreateWithId creates a recipe with explicit ID
func (f *RecipeFactory) CreateWithId(
	id valueobject.Id,
	name string,
	ingredients []valueobject.RecipeIngredient,
	totalCost valueobject.Money,
) (*entity.Recipe, error) {
	return entity.NewRecipe(
		id,
		name,
		ingredients,
		totalCost,
	)
}